src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+        private const string ec2BaseUri = "http://169.254.169.254";
+        private const int randomLength = 32;
+
+        // static properties
+        /// <summary>
+        /// Gets the name of the mechanism.
+        /// </summary>
+        /// <value>
+        /// The name of the mechanism.
+        /// </value>
+        public static string MechanismName
+        {
+            get { return "MONGODB-AWS"; }
+        }
+
+        // constructors
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GssapiAuthenticator"/> class.
+        /// </summary>
+        /// <param name="credential">The credentials.</param>
+        /// <param name="properties">The properties.</param>
+        public MongoAWSAuthenticator(UsernamePasswordCredential credential, IEnumerable<KeyValuePair<string, string>> properties)
+            : base(CreateMechanism(credential, properties, new DefaultRandomByteGenerator()))
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="GssapiAuthenticator"/> class.
+        /// </summary>
+        /// <param name="username">The username.</param>
+        /// <param name="properties">The properties.</param>
+        public MongoAWSAuthenticator(string username, IEnumerable<KeyValuePair<string, string>> properties)
+            : base(CreateMechanism(username, null, properties, new DefaultRandomByteGenerator()))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGO AWS authentication may only use the $external source.", "credential");
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator)
+        {
+            var relativeUri = Environment.GetEnvironmentVariable(awsContainerCredentialsRelativeUriKey);
+            var methodsList = new Func<AwsCredentials>[]
+            {
+                () => AcquireCredentialsFromMongoCredentials(username, securePassword, properties),
+                () => AcquireMechanismDataFromEnvironmentVariables(),
+                () => AcquireMechanismDataFromEcsResponse(ecsBaseUri, relativeUri),
+                () => AcquireMechanismDataFromEc2Response(ec2BaseUri)
```

Comment:
Obtaining credentials flow is described in specification: [link](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#obtaining-credentials)